io-sim-classes/src/Control/Monad/Class/MonadThrow.hs

```diff
@@ -92,11 +98,26 @@ class MonadThrow m => MonadCatch m where
               _ <- what
               throwM (e :: SomeException)
 
-  bracketOnError before after thing =
-    mask $ \restore -> do
-      a <- before
-      restore (thing a) `onException` after a
+  generalBracket acquire release use =
+    mask $ \unmasked -> do
+      resource <- acquire
+      b <- unmasked (use resource) `catch` \e -> do
+        _ <- release resource (ExitCaseException e)
+        throwM e
+      c <- release resource (ExitCaseSuccess b)
+      return (b, c)
 
+-- | Perform final action only if an error is thrown
+--
+-- Implementation in terms of 'generalBracket' as in @exceptions@
+bracketOnError :: MonadCatch m => m a -> (a -> m b) -> (a -> m c) -> m c
+bracketOnError acquire release = liftM fst . generalBracket
```

Comment:
I think we should stick to the approach we follow in the other functions that our IO instances are exactly the IO functions, so including these utils into the type class, rather than redefining them in terms of the class primitives.